fabric-transfer-api-v1/src/main/java/net/fabricmc/fabric/api/transfer/v1/storage/Storage.java

```diff
@@ -117,6 +117,20 @@ default boolean supportsExtraction() {
 	 */
 	Iterator<StorageView<T>> iterator(Transaction transaction);
 
+	/**
+	 * Iterate through the contents of this storage, for the scope of the passed transaction.
+	 * This function follows the semantics of {@link #iterator}, but returns an {@code Iterable} for use in {@code for} loops.
+	 *
+	 * @param transaction The transaction to which the scope of the returned iterator is tied.
+	 * @return An iterator over the contents of this storage.
+	 * @throws IllegalStateException If this storage is already exposing storage views.
+	 * @see #iterator
+	 */
+	default Iterable<StorageView<T>> iterable(Transaction transaction) {
+		Iterator<StorageView<T>> iterator = iterator(transaction);
+		return () -> iterator;
```

Comment:
should you call the iterator method upon this method call or only when the iterable is used?

Reply:
```suggestion
		return () -> iterator(transaction);
```

Reply:
Iteration is mutually exclusive, so I want this "locking" to happen directly.

Reply:
(I think I saw the message 2 days ago, but forgot to answer... my bad)

Reply:
what

Reply:
Can you be more precise? What is confusing you? The fact that iteration has mutex semantics, or the fact that I chose to acquire it directly instead of in the callback?

Reply:
it will, but that's the job of `iterator`.

Reply:
the lock is closed when the transaction is ;)

Reply:
"it will, but that's the job of the iterator." the exception thrown will be unintuitive and may not be thrown all the time if the iterable is not implemented normally. eg. instead of "reached end of inventory" which tells you nothing, you would have "cannot use iterable twice"

"the lock is closed when the transaction is ;)" that means you can only use the iterable once per transaction, that doesn't make sense

Reply:
https://github.com/Technici4n/fabric/blob/8c8e128c7d4d5d0ef1c52af5175299fe0c2d875e/fabric-transfer-api-v1/src/main/java/net/fabricmc/fabric/api/transfer/v1/storage/base/CombinedStorage.java#L85
I think the exception is clear enough.

Reply:
> "the lock is closed when the transaction is ;)" that means you can only use the iterable once per transaction, that doesn't make sense

Transactions are the only scope we have I think. We don't really have a choice... They are quite cheap, so you can just do your iteration in a nested transaction in case you need to do multiple.

Reply:
But why, it's an arbitrary restriction, inconvenient and unintuitive. Iterable is Iterable, not a wrapper around Iterator `List` has the exact same contract or whatever and it doesn't need to do this

Reply:
It's designed that way to allow more flexibility for dynamic inventories, for example inventories that can contain up to 1k items, regardless of their type.

Reply:
why can't that be done normally?

Reply:
I'm trying to find the reason... (I may be overthinking this?)

Reply:
I think the idea was that if you have some kind of dynamic collection (for example hashmap) you want to know which iterators are in scope so that you can update them.

Reply:
But that would work with multiple iterators too I think?

Reply:
as of right now there can only be one, and yea it wouldn't be much harder to make it work with multiple iterators

Reply:
That should work, I'll update the PR with this soonish.

Reply:
Updated.